raster-foundry/raster-foundry
# Add workflow for releasing to Maven Central

## Overview

* Add back support for [Coursier](https://github.com/coursier/coursier) for parallel artifact fetching
* Add [sbt-release](https://www.github.com/sbt/sbt-release) plugin
* Move publishing configuration to `build.sbt`
* Add `noPublishSettings` construct to so we can run `$ sbt release` on the root project without building projects like `backsplash` and `geotrellis`
* Refactor `cipublish` to take advantage of `noPublishSettings` and run `$ sbt publishSigned`
* Add Maven Central release pipeline on Jenkins
* Add `ValidatingStringParameterDefinition` plugin
* Refactor `bootstrap` script
* Update `cibuild` to reference refactored `bootstrap` script

Closes #3793 

## Notes

The scope of the original issue was based on my understanding that we could promote artifacts that were already published as snapshots to release artifacts, similar to how we retag container images published from `develop` in `Jenkinsfile.release`. This turned out to be wrong.

Instead, there are two separate repositories: one for snapshots, and one for staging release artifacts. The presence of a `-SNAPSHOT` qualifier in the sbt project version is what determines which repository the artifact is published to. You can think of `$ sbt release` as a wrapper around the command we're already using to publish snapshots that strips the qualifier.

The addition of a new pipeline for releasing artifacts introduces an inconsistency that a platform release of a same version has an assembly jar that has different version metadata than the Maven Central release of a same version. This is because the container images are built in the `develop` pipeline where the version has the `-SNAPSHOT` qualifier.

We're overloading the notion of versioning, because we’re tagging the artifacts with X version and releasing the platform at a different point of time.

To do this correctly would introduce a lot of complications to the build pipeline. I think it would be valuable to have a discussion around this.

## Testing Instructions

Follow the [wiki instructions](https://github.com/azavea/raster-foundry-deployment/wiki/Release-Workflow#publish-jars-to-maven-central) to start a new Maven Central release pipeline.

See: 

http://jenkins.staging.rasterfoundry.com/job/Raster%20Foundry%20Maven%20Central%20Release%20Pipeline/19/console
